### ISSUE TITLE
FAQ - Why is server authentication relevant if we have the repo locally?

### ISSUE THREAD

**Issue by Korporal** [GitTools/GitLink#147]
I notice you point out that private Github repos are a concern because authentication is often required. But if the repo is local (as it almost always is for developers) then the source code is also present locally for any given commit ID.

Therefore why is it not possible to bypass Github entirely and use the local (cloned) repo?

Thanks

PS: I understand now what you're referring to (shared external packages) but wanted to get more info about how to set things up when the use case is primarily debugging one's own packages who's source is stored in one's own private Github repos.

**Comment by GeertvanHorrik** [comment]
I'll first explain how GitLink works.

1. It opens your pdb file and checks all the local files (in this case the build agent or your local machine)
2. Then it checks the .git directory and checks the remotes for these files. This allows the tooling to map local files => remotely hosted files
3. It updates the pdb

When debugging in Visual Studio (and the pdb is available), Visual Studio will remap the files in the pdb (which are always local files) to the remote ones.

Now the reason you (sometimes) need authentication is because of private repositories. But since Visual Studio doesn't really support authentication for source servers, it's a bit cumbersome to use. I believe @ctaggart created a solution (I think some sort of proxy) to bypass this.

**Comment by Korporal** [comment]
Thanks @GeertvanHorrik for explaining this.

A question then, would it be feasible to add some feature that allows us to refer to our local (cloned) repo rather than our Github server? This is what I meant by an alternative/simpler process - one in which we know ahead of time that the source code will always be inside our own repos. Because we know its always inside our own repos the mechanism could be changed in some way to not refer to Github but to the local repo - since we know we have our own source code in there?

**Comment by GeertvanHorrik** [comment]
That would be hard because it needs to map from _x_ => _y_. 

x => the *original build source file location*
y => the source file on the end-user running the dependency

The issue is that you don't know _y_, so this must either be exactly the same for everyone, or you must research whether you can remap this dynamically. And even then, you need to point to a specific point in time (_x_ might be built with a different commit than _y_). When the checksums don't match, you will have an issue and this wouldn't work.

Long story short: I would go for trying to use the proxy that @ctaggart wrote or use the powershell authentication (which I don't know too much about).

**Comment by ctaggart** [comment]
@Korporal, do you have a private network? You could easily create a proxy for https://raw.githubusercontent.com/. That is what I did here: https://github.com/ctaggart/SourceLink-proxy. 

If you plan to adopt the new dotnet tooling, you can also just embed the source files in the Portable PDB files. If you don't like the idea of shipping source code around, then update 1 for Visual Studio 2017 will probably have private repository support. https://github.com/dotnet/roslyn/issues/12759#issuecomment-282793617

**Comment by GeertvanHorrik** [comment]
Closing ticket. Feel free to re-open if not answered.
